src/HealthChecks/HealthChecks/src/HealthCheckPublisherHostedService.cs

```diff
@@ -104,21 +105,26 @@ private async void Timer_Tick(object state)
             await RunAsync();
         }
 
+        // Internal for testing
+        internal void CancelToken()
+        {
+            _runTokenSource.Cancel();
+        }
+
         // Internal for testing
         internal async Task RunAsync()
         {
             var duration = ValueStopwatch.StartNew();
             Logger.HealthCheckPublisherProcessingBegin(_logger);
 
-            CancellationTokenSource cancellation = null;
             try
             {
                 var timeout = _options.Value.Timeout;
 
-                cancellation = CancellationTokenSource.CreateLinkedTokenSource(_stopping.Token);
-                cancellation.CancelAfter(timeout);
+                _runTokenSource = CancellationTokenSource.CreateLinkedTokenSource(_stopping.Token);
```

Comment:
Can you keep the `cancellation` local variable and just *assign* both `cancellation` and `_runTokenSource` here?

The problem I'm worried about is the other effects of turning this into a field from a local. There's nothing in this code that prevents `Timer_Tick` from overlapping. In the event that happens, this logic is all messed up, and would probably result in leaking a CTS (not disposed).

So this would still be pretty easy to reason about if the change **just** adds `_runTokenSource` as a thing that's used in tests, and the main code path stuff referred to the token through `cancellation`.

Reply:
Yeah, I was wondering about timer overlaps and almost did that :+1: